src/PasswordHasher/DefaultPasswordHasher.php

```diff
@@ -40,10 +40,10 @@ class DefaultPasswordHasher extends AbstractPasswordHasher
     /**
      * Generates password hash.
      *
-     * @param string $password Plain text password to hash.
+     * @param array|string $password Plain text password to hash.
      * @return string Password hash or false on failure.
      */
-    public function hash($password): string
+    public function hash(array|string $password): string
```

Comment:
PasswordHasherIdentifier requires `array|string` for password, but none of the hashers support arrays.

Reply:
That doesn't sound right. Where exactly does the identifier pass an array for the password?

Reply:
It was just part of the @param annotation that was converted into a type declaration.

Reply:
Sorry, I still don't see it. Which `@param`?

Reply:
https://github.com/cakephp/authentication/blob/cake5/src/PasswordHasher/PasswordHasherInterface.php#L26

Reply:
I can't remember either why it's `string|array`. Maybe to support some edge case custom hashing is my guess.

Reply:
Ah OK, you meant the interface, I thought the typo was meant to be `PasswordIdentifier`.

Reply:
Maybe this is a chance to drop the array? With parameter type widening, one could implement a `string` only param as `array|string`, or even drop the type hint completely if needed.

The password identifier doesn't support passing arrays, so a custom identifier would be needed anyways if people wanted to pass arrays to the hasher, so they could then also "cast" the used hasher accordingly.

Reply:
So we want to drop `array`? That would fix the mismatch between needsRehash() and hash().

Reply:
I don't really see a mismatch there, as `needsRehash()` operates on the hashed password (maybe the argument should be renamed to reflect that better). What makes me feel uneasy is forcing all the hashers to accept an array when they don't support it, and will fail with type errors accordingly 🤷

Reply:
ah, you're right, it takes the hashed password. I must have been looking at a different password_* parameter.